_posts/2015-04-24-FFI-and-Rust.md

```diff
@@ -0,0 +1,264 @@
+verify this claim of zero overhead. Almost all programming languages can call
+into C one way or another, but it often comes at a cost with runtime type
+coversions or perhaps some language runtime juggling. To get a handle on what
+Rust is doing, let's go straight to the assembly code of the above `main`
+function:
+
+```
+mov    $0x4,%edi
+callq  3bc30 <double_input>
+```
+
+And as before, that's it! Here we can see that calling a C function from Rust
+involves precisely one call instruction after moving the arguments into place,
+exactly the same cost as it would be in C.
+
+### Safe Abstractions
+
+One of Rust's core design principles is its emphasis on ownership, and FFI is no
+exception here. When binding a C library in Rust you not only have the benefit
+of 0 overhead, but you are also able to make it *safer* than C can! Bindings
+can leverage the ownership and borrowing principles in Rust to codify comments
+typically found in a C header about how its API should be used.
+
+For example, consider a C library for parsing a tarball. This library will
+expose functions to read the contents of each file in the tarball, probably
+something along the lines of:
+
+```c
+// Gets the data for a file in the tarball at the given index, returning NULL if
+// it does not exist. The `size` pointer is filled in with the size of the file
+// if successful.
+const char *tarball_file_data(tarball_t *tarball, unsigned index, size_t *size);
+```
+
+This function is implicitly making assumptions about how it can be used,
+however, by assuming that the `char*` pointer returned cannot outlive the input
+tarball. When bound in Rust, this API might look like this instead:
+
+```rust
+pub struct Tarball { raw: *mut tarball_t }
+
+impl Tarball {
+    pub fn file(&self, index: u32) -> Option<&[u8]> {
+        unsafe {
+            let mut size = 0;
+            let data = tarball_file_data(self.raw, index as libc::c_uint,
+                                         &mut size);
+            if data.is_null() {
+                None
+            } else {
+                Some(slice::from_raw_parts(data as *const u8, size as usize))
+            }
+        }
+    }
+}
+```
+
+Here the `*mut tarball_t` pointer is *owned by* a `Tarball`, so we already have
+rich knowledge about the lifetime of the resource. Additionally, the `file`
+method returns a **borrowed slice** whose lifetime is connected to the same
+lifetime as the source tarball itself. This is Rust's way of indicating that the
+returned data cannot outlive the tarball, statically preventing bugs that may be
+encountered when just using C.
+
+A key aspect of the Rust binding here is that it is a safe function! Although it
+has an `unsafe` implementation (due to calling an FFI function), this interface
+is safe to call and will not cause tough-to-track-down segfaults. And don't
+forget, all of this is coming at 0 cost as the raw types in C are representable
+in Rust with no extra allocations or overhead.
```

Comment:
This safe interface imposes an extra branch over using the raw interface because it checks for a null return and the caller is going to have to duplicate that check of the Option.